Add support for running a local DNS resolver for WPT

annevk opened the pull request:
#### 060bebfb1e6c057745cfc21a69433d5d9ea5c541
<pre>
Add support for running a local DNS resolver for WPT
<a href="https://bugs.webkit.org/show_bug.cgi?id=261038">https://bugs.webkit.org/show_bug.cgi?id=261038</a>
<a href="https://rdar.apple.com/82927182">rdar://82927182</a>

Reviewed by NOBODY (OOPS!).

Some ports don&apos;t have support for resolving arbitrary domains, but
web-platform-tests rely on this capability. Currently, the Apple ports only
use localhost, and 127.0.0.1 as a workaround where it&apos;s necessary, but this
affects our test coverage. This patch adds a very basic DNS resolver and
introduces a new command-line argument for run-webkit-tests; this is disabled
by default.

In 254273@main, glib ports gained support for resolving arbitrary domains,
and this moves Apple ports closer to supporting that. We expect to enable
this by default, along with rebaselining, in a later commit.

Previous attempts were made in <a href="https://bugs.webkit.org/show_bug.cgi?id=245294">https://bugs.webkit.org/show_bug.cgi?id=245294</a>,
and 264076@main introduced part of that behavior for a different reason. In
some situations, name resolutions falls back on that SPI, as well.

The basic DNS resolver only resolves A records for names ending with
&quot;localhost&quot; and for the host names defines by the port&apos;s localhost_aliases.
The server always returns 127.0.0.1 as the answer for those hosts. All other
queries return NXDomain. This resolver depends on the dnslib Python library.

On Apple ports, the iOS Simulator and Mac use different implementations. The
iOS simulator part is in the Network process because we need to define the
resolver in-process otherwise it won&apos;t be used. The way the resolver is
configured on Mac isn&apos;t supported on the simulator. On Mac, we use SPI and a
Network Extensions policy so this works when a VPN is enabled. We do this in
WebKitTestRunner so we don&apos;t need to add additional entitlements and weaken
the sandbox for the network process.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/060bebfb1e6c057745cfc21a69433d5d9ea5c541

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108656 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28317 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113867 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59041 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110619 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29006 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36873 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82542 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111604 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23035 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97876 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62979 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/108088 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22453 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16013 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58574 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92405 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16064 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116988 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35710 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26341 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91561 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36083 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94145 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91365 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36268 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14028 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31542 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35611 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41144 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35321 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38667 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36998 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/port/test.py:
```diff
@@ -30,7 +30,8 @@
 
-from webkitpy.port import Port, Driver, DriverOutput
+from webkitpy.port.base import Port
+from webkitpy.port import Driver, DriverOutput
```
```suggestion
from webkitpy.port.driver import Driver, DriverOutput
```

gsnedders commented on Tools/Scripts/webkitpy/port/__init__.py:
```diff
@@ -28,6 +28,5 @@
 
 """Port-specific entrypoints for the layout tests test infrastructure."""
 
-from webkitpy.port.base import Port  # It's possible we don't need to export this virtual baseclass outside the module.
```
This is almost certainly relied on all over the place, and we won't be able to get away with this I'm pretty sure?

gsnedders commented on Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
```diff
@@ -116,6 +118,18 @@ static void initializeNetworkSettings()
 #endif
     m_enableModernDownloadProgress = parameters.enableModernDownloadProgress;
+
+#if PLATFORM(IOS_FAMILY_SIMULATOR)
```
Should this not be `PLATFORM(IOS_FAMILY)`? Not sure why we'd only apply this to Simulator.

annevk replied:
I'm not sure either. @sysrqb could you comment on this? I suppose I'll try to see what happens if I change this to `IOS_FAMILY`.

sysrqb replied:
I didn't test it on iOS. It's worth supporting, but is it a requirement?

gsnedders commented on Tools/Scripts/webkitpy/__init__.py:
```diff
@@ -78,6 +78,7 @@
 AutoInstall.register(Package('configparser', Version(4, 0, 2), implicit_deps=['pyparsing'], aliases=['backports.configparser']))
 AutoInstall.register(Package('contextlib2', Version(0, 6, 0)))
 AutoInstall.register(Package('coverage', Version(7, 6, 1), wheel=True))
+AutoInstall.register(Package('dnslib', Version(0, 9, 23)))
```
```suggestion
AutoInstall.register(Package('dnslib', Version(0, 9, 26)))
```

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/servers/basic_dns_server.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class Resolver(BaseResolver):
+    def __init__(self, allowed_hosts=[]):
```
Objects are only initialised once when they appear in the parameter spec, so it's the same list object for all calls, which we probably don't want (though I realise in this case we don't actually modify it ever).

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/servers/basic_dns_server.py:
```diff
@@ -0,0 +1,57 @@
+_log = logging.getLogger(__name__)
+
+
+class Resolver(BaseResolver):
```
I have a few unposted minor comments on the implementation of this, but it's probably not worthwhile spending too much time on — once https://github.com/web-platform-tests/wpt/pull/52416 lands and we reimport the tools we can just use the upstream DNS server.

sysrqb replied:
@gsnedders Are you recommending that this patch should depend on https://github.com/web-platform-tests/wpt/pull/52416?

gsnedders replied:
No strong preference; happy for us to land this and then migrate later.

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/servers/web_platform_test_server.py:
```diff
@@ -127,6 +128,10 @@ def __init__(self, port_obj, name, pidfile=None):
         self._process = None
+        self._dns_server = None
+        if port_obj.supports_localhost_aliases and port_obj.get_option('local_dns_resolver') and not port_obj.get_option('disable_wpt_hostname_aliases'):
+            self._dns_server = DNSServer(Resolver(allowed_hosts=port_obj.localhost_aliases()), port=8053, address="127.0.0.1")
```
While I realise there are other very long lines in this file, can we wrap these to a more normal line length?

annevk replied:
`autopep8` didn't wrap the if statement...

gsnedders replied:
`autopep8` probably won't, because you can't wrap it unless you add parenthesis around it, and `autopep8` exclusively makes whitespace changes.

gsnedders commented on Tools/Scripts/webkitpy/port/apple.py:
```diff
@@ -83,6 +83,7 @@ def __init__(self, host, port_name, **kwargs):
 
         port_name = port_name.replace('-wk2', '')
         self._version = self._strip_port_name_prefix(port_name)
+        self.supports_localhost_aliases = True
```
```suggestion
        self.supports_localhost_aliases = False
```

if the intention is still to land this still disabled?

gsnedders commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -232,6 +234,42 @@ void initializeWebViewConfiguration(const char* libraryPath, WKStringRef injecte
+    [policySession.get() addPolicy:agentPolicy];
+
+    policySession.get().priority = NEPolicySessionPriorityHigh;
+    [policySession.get() apply];
```
If we want to support iOS (non-Simulator), do we need the same VPN supporting code there?

sysrqb replied:
Do we currently support running WPT on iOS devices?

gsnedders replied:
We have support for this via the `Tools/Scripts/webkitpy/port/ios_device.py`, but it's definitely lower priority than getting Simulator working.

gsnedders reviewed (changes requested):
Per below, my main questions are about "why are we limiting this to macOS and iOS family Simulator, why not macOS and iOS family"? (cc @bls1999)

I do think the commit message could be clarified too, maybe something like:

```diff
diff --git a/a b/b
index 376cde4..09ca778 100644
--- a/a
+++ b/b
@@ -4,15 +4,15 @@ rdar://82927182
 
 Reviewed by NOBODY (OOPS!).
 
-Some ports don't have support for resolving arbitrary domains, but Web Platform
-Tests rely on this capability. Currently, we only use localhost and 127.0.0.1
+Some ports don't have support for resolving arbitrary domains, but
+web-platform-tests rely on this capability. Currently, the Apple ports only use localhost, and 127.0.0.1
 as a workaround where it's necessary, but this affects our test coverage. This
 patch adds a very basic DNS resolver and introduces a new command-line argument
-for run-webkit-tests. Disabled by default so it can be enabled along with
-rebased expectations separately.
+for run-webkit-tests; this is disabled by default.
 
 In 254273@main, glib ports gained support for resolving arbitrary domains, and
-this moves Apple ports closer to supporting that.
+this moves Apple ports closer to supporting that. We expect to enable this by default, along with
+rebaselining, in a later commit.
 
 Previous attempts were made in https://bugs.webkit.org/show_bug.cgi?id=245294,
 and 264076@main introduced part of that behavior for a different reason. In
@@ -21,9 +21,9 @@ some situations, name resolutions falls back on that SPI, as well.
 The basic DNS resolver only resolves A records for names ending with
 "localhost" and for the host names defines by the port's localhost_aliases.
 The server always returns 127.0.0.1 as the answer for those hosts. All other
-queries return NXDomain. This resolver depends on the dnslib python library.
+queries return NXDomain. This resolver depends on the dnslib Python library.
 
-On Apple ports, the iOS simulator and Mac use different implementations. The
+On Apple ports, the iOS Simulator and Mac use different implementations. The
 iOS simulator part is in the Network process because we need to define the
 resolver in-process otherwise it won't be used. The way the resolver is
 configured on Mac isn't supported on the simulator. On Mac, we use SPI and a
```

Also: the `no-failure-limits` label may be useful if you _do_ want to try running it this PR again, with it still enabled (despite the commit message)?

achristensen07 commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -232,6 +234,42 @@ void initializeWebViewConfiguration(const char* libraryPath, WKStringRef injecte
+
+    // The following NetworkExtension policy is needed so we can run tests while a VPN is connected
+    static RetainPtr<NEPolicySession> policySession = adoptNS([[NEPolicySession alloc] init]);
+    NEPolicy *domainPolicy = [[NEPolicy alloc] initWithOrder:1 result:[NEPolicyResult netAgentUUID:agentUUID.get()] conditions:@[ [NEPolicyCondition domain:@"test"] ]];
```
This needs adoptNS or it's a memory leak.

achristensen07 commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -232,6 +234,42 @@ void initializeWebViewConfiguration(const char* libraryPath, WKStringRef injecte
+    [policySession.get() addPolicy:domainPolicy];
+
+    NSString *agentString = agentUUID.get().UUIDString;
+    NEPolicy *agentPolicy = [[NEPolicy alloc] initWithOrder:1 result:[NEPolicyResult netAgentUUID:agentUUID.get()] conditions:@[ [NEPolicyCondition domain:agentString] ]];
```
Ditto.

achristensen07 commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -232,6 +234,42 @@ void initializeWebViewConfiguration(const char* libraryPath, WKStringRef injecte
+    }
+
+    // The following NetworkExtension policy is needed so we can run tests while a VPN is connected
+    static RetainPtr<NEPolicySession> policySession = adoptNS([[NEPolicySession alloc] init]);
```
Since it seems like this code should only be done once per process launch instead of once per call to initializeWebViewConfiguration, we should probably early return if we've already called NEPolicySession. apply.
Also, this would be good to put in its own function.

sysrqb committed:
Add support for running a local DNS resolver for WPT
https://bugs.webkit.org/show_bug.cgi?id=261038
rdar://82927182

Reviewed by NOBODY (OOPS!).

Some ports don't have support for resolving arbitrary domains, but
web-platform-tests rely on this capability. Currently, the Apple ports only
use localhost, and 127.0.0.1 as a workaround where it's necessary, but this
affects our test coverage. This patch adds a very basic DNS resolver and
introduces a new command-line argument for run-webkit-tests; this is disabled
by default.

In 254273@main, glib ports gained support for resolving arbitrary domains,
and this moves Apple ports closer to supporting that. We expect to enable
this by default, along with rebaselining, in a later commit.

Previous attempts were made in https://bugs.webkit.org/show_bug.cgi?id=245294,
and 264076@main introduced part of that behavior for a different reason. In
some situations, name resolutions falls back on that SPI, as well.

The basic DNS resolver only resolves A records for names ending with
"localhost" and for the host names defines by the port's localhost_aliases.
The server always returns 127.0.0.1 as the answer for those hosts. All other
queries return NXDomain. This resolver depends on the dnslib Python library.

On Apple ports, the iOS Simulator and Mac use different implementations. The
iOS simulator part is in the Network process because we need to define the
resolver in-process otherwise it won't be used. The way the resolver is
configured on Mac isn't supported on the simulator. On Mac, we use SPI and a
Network Extensions policy so this works when a VPN is enabled. We do this in
WebKitTestRunner so we don't need to add additional entitlements and weaken
the sandbox for the network process.
